File: DAA/Lab 3/Fractional Knapsack.py
def fractional_knapsack(weights, values, capacity):
    if len(weights) != len(values):
        raise ValueError("Weights and values must be of the same length.")
    res = 0
    for pair in sorted(zip(weights, values), key=lambda x: x[1] / x[0], reverse=True):
        if capacity <= 0:
            break
        if pair[0] > capacity:
            res += capacity * (pair[1] / pair[0])
            capacity = 0
        else:
            res += pair[1]
            capacity -= pair[0]
    return res
def main():
    num_items = int(input("Enter the number of items: "))
    weights = list(map(float, input("Enter the weights (space-separated): ").split()))
    values = list(map(float, input("Enter the values (space-separated): ").split()))
    if len(weights) != num_items or len(values) != num_items:
        raise ValueError("The number of weights and values must match the number of items.")
    capacity = float(input("Enter the capacity of the knapsack: "))
    max_value = fractional_knapsack(weights, values, capacity)
    print(f"Maximum value in the knapsack: {max_value}")
if __name__ == "__main__":
    main()

File: DAA/Lab 3/FractionalKnapsack.java
import java.util.*;
public class FractionalKnapsack {
    public static double fractionalKnapsack(double[] weights, double[] values, double capacity) {
        if (weights.length != values.length) {
            throw new IllegalArgumentException("Weights and values must be of the same length.");
        }
        double res = 0;
        List<Item> items = new ArrayList<>();
        for (int i = 0; i < weights.length; i++) {
            items.add(new Item(weights[i], values[i]));
        }
        items.sort((a, b) -> Double.compare(b.value / b.weight, a.value / a.weight));
        for (Item item : items) {
            if (capacity <= 0) {
                break;
            }
            if (item.weight > capacity) {
                res += capacity * (item.value / item.weight);
                capacity = 0;
            } else {
                res += item.value;
                capacity -= item.weight;
            }
        }
        return res;
    }
    private static class Item {
        double weight;
        double value;
        Item(double weight, double value) {
            this.weight = weight;
            this.value = value;
        }
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter the number of items: ");
        int numItems = scanner.nextInt();
        double[] weights = new double[numItems];
        System.out.print("Enter the weights (space-separated): ");
        for (int i = 0; i < numItems; i++) {
            weights[i] = scanner.nextDouble();
        }
        double[] values = new double[numItems];
        System.out.print("Enter the values (space-separated): ");
        for (int i = 0; i < numItems; i++) {
            values[i] = scanner.nextDouble();
        }
        if (weights.length != numItems || values.length != numItems) {
            throw new IllegalArgumentException("The number of weights and values must match the number of items.");
        }
        System.out.print("Enter the capacity of the knapsack: ");
        double capacity = scanner.nextDouble();
        double maxValue = fractionalKnapsack(weights, values, capacity);
        System.out.printf("Maximum value in the knapsack: %.2f%n", maxValue);
        scanner.close();
    }
}
